# basic numpy arrray 

### creating numpy  array

In [1]:
import numpy as np

In [2]:
a=np.array([1,2,3]) #creating numpy array
print(type(a)) #type show data of a veriable

<class 'numpy.ndarray'>


In [3]:
b=np.array([[1,2,3],[4,5,6]])
print(b)#creating a 2d array

[[1 2 3]
 [4 5 6]]


In [4]:
c=np.array([[[1,2,3],[4,5,6]],[[7,8,9],[10,11,12]]])
print(c)

[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]


### to set or change dtype

In [5]:
np.array([1,2,3],dtype=float)#it make all value float

array([1., 2., 3.])

In [6]:
np.array([1,2,3,0],dtype=bool)#it show True for all values but it False for only 0 

array([ True,  True,  True, False])

explaination of above code 
dtype=bool tells NumPy to convert each element to a Boolean (True or False).
In Python/NumPy:
0 → False
Any non-zero number → True

In [7]:
np.array([1,2,3],dtype=complex)

array([1.+0.j, 2.+0.j, 3.+0.j])

🔹 Explanation:
dtype=complex tells NumPy to store each element as a complex number.
A complex number has the form a + bj, where:
a = real part
b = imaginary part
It converts each integer into a complex number with an imaginary part of 0j.

### How to use arange

In [8]:
np.arange(1,11) #

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

🔹 Explanation:
np.arange(start, stop) creates an array with numbers starting from start up to stop (but not including stop).
🔹 Here:
start = 1
stop = 11
So it gives numbers from 1 to 10.
Creates a NumPy array of numbers from 1 to 10 (stop value 11 is excluded).
and it works similarly like range 

### How to reshape and array

🔹 reshape() function (NumPy)
Used to change the shape of an array without changing its data.
✅ In short:
reshape() rearranges the same data into a new shape (rows, columns, or dimensions).

In [9]:
np.arange(1,13).reshape(2,6)#it an array of 1d by default this function help us it make it 2D or 3D.
#this array is 2D

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12]])

In [10]:
np.arange(1,13).reshape(2,3,2)
#make an 3d array

array([[[ 1,  2],
        [ 3,  4],
        [ 5,  6]],

       [[ 7,  8],
        [ 9, 10],
        [11, 12]]])

maximum time arange and reshape function will be used together 

### How to use function ones and zeros

🔹 np.ones((3,4))
Creates a 3×4 array filled with 1s.
Useful when you need an array of all ones — for example, as a starting point for calculations or masking.
🔹 np.zeros((3,4,2))
Creates a 3×4×2 array filled with 0s.
Useful for initializing empty data structures, placeholders, or starting arrays for computation.
format (dimention,rows,colums)

In [11]:
np.ones((3,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [12]:
np.zeros((3,4,2))

array([[[0., 0.],
        [0., 0.],
        [0., 0.],
        [0., 0.]],

       [[0., 0.],
        [0., 0.],
        [0., 0.],
        [0., 0.]],

       [[0., 0.],
        [0., 0.],
        [0., 0.],
        [0., 0.]]])

### How to use random function

🔹 np.random.random((3,4,2))
format in(dimention,rows,colums)
Creates a 3×4×2 array of random float numbers between 0.0 and 1.0.
➡️ Think of it as 3 blocks, each with 4 rows and 2 random values per row.
🔹 np.random.random((3,4))
Creates a 3×4 matrix (2D array) of random floats between 0.0 and 1.0.
➡️ 3 rows × 4 columns of random numbers.
🔹 np.random.random(3)
Creates a 1D array with 3 random numbers between 0.0 and 1.0.
➡️ Something like [0.12, 0.78, 0.45]

🧠 In short:
All generate random floats in [0, 1).
The numbers inside parentheses define the shape (dimensions) of the array.



In [13]:
np.random.random((3,4,2)) #it gives 3D array of random numbers

array([[[0.03410605, 0.21368593],
        [0.45046684, 0.16906775],
        [0.1570719 , 0.50397867],
        [0.76490163, 0.43313584]],

       [[0.18073693, 0.45164397],
        [0.21997596, 0.82986668],
        [0.73594449, 0.52560613],
        [0.15611255, 0.22582241]],

       [[0.74826651, 0.1837627 ],
        [0.93466307, 0.73411609],
        [0.0572082 , 0.30787247],
        [0.35728636, 0.43075984]]])

In [14]:
np.random.random((3,4))#it gives 2 d array of random numbers

array([[0.75312859, 0.98077861, 0.8955331 , 0.15277014],
       [0.57305082, 0.27210565, 0.28664315, 0.58260463],
       [0.05361037, 0.25821248, 0.37053081, 0.87441543]])

In [15]:
np.random.random(3)#it gives 1d array of randon 

array([0.61122726, 0.13059226, 0.2010174 ])

### How to use linspace

🔹 What it does
np.linspace() creates a sequence of evenly spaced numbers between a start and end value.
🔹 Syntax
np.linspace(start, stop, num)
start → starting value
stop → ending value
num → number of values you want between start and stop (default = 50)
🔹 Explanation
It divided the range from 1 to 10 into 5 equal parts (including both ends).
🧠 In short:
np.linspace() → gives equally spaced values between two numbers — useful in graphs, data sampling, or plotting.

In [16]:
np.linspace(-10,10,10)

array([-10.        ,  -7.77777778,  -5.55555556,  -3.33333333,
        -1.11111111,   1.11111111,   3.33333333,   5.55555556,
         7.77777778,  10.        ])

In [17]:
np.linspace(1,11,10)# 11 is include in this function

array([ 1.        ,  2.11111111,  3.22222222,  4.33333333,  5.44444444,
        6.55555556,  7.66666667,  8.77777778,  9.88888889, 11.        ])

In [18]:
np.linspace(1,10,10)

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

### How to make identity matrix

you can create an identity matrix in NumPy using the np.identity() function.
🔹 Syntax
np.identity(n)
n → number of rows (and columns, since it’s a square matrix)
🔹 Explanation
✅ It creates a square matrix with
1s on the main diagonal
0s everywhere else
🧠 In short:
np.identity(n) → makes an n×n identity matrix, often used in linear algebra (like matrix multiplication or inversion).

In [19]:
import numpy as np
I = np.identity(3)
I[I == 0] = 5   # replace all 0s with 5
print(I)

[[1. 5. 5.]
 [5. 1. 5.]
 [5. 5. 1.]]


In [20]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [21]:
a1=np.arange(10)
a2=np.arange(12,dtype=float).reshape(3,4)
a3=np.arange(8).reshape(2,2,2)

### How to use ndim

✅ What is ndim?
ndim is an attribute of a NumPy array.
It returns an integer showing how many dimensions the array has.
✅ Why is ndim important?
Understanding dimensions helps you:
Use reshape()
Work with matrices
Prepare data for ML models (which need 2D or 3D arrays)
Use operations like sum(axis=...), flatten(), etc.

In [22]:
a3.ndim #output is 3 because it is 3Darray

3

In [23]:
a2.ndim

2

In [24]:
a1.ndim

1

### Shape function

✅ What is shape?
shape returns a tuple that shows how many elements exist in each dimension of the array.
✅ shape tells you how many elements are present in each dimension.
✅ Very useful for reshaping arrays, matrix operations, ML models, etc.

In [25]:
arr = np.array([[1,2],[3,4],[5,6]])
print(arr.shape) #(rows,colunms)


(3, 2)


In [26]:
data = np.array([[1,2,3],[4,5,6]])
rows, cols = data.shape
print(rows, cols)


2 3


In [27]:
a1.shape

(10,)

In [28]:
a2.shape

(3, 4)

In [29]:
a3.shape

(2, 2, 2)

### What is item size and size

✅ 1. What is itemsize in NumPy?
itemsize tells you how many BYTES each element in a NumPy array uses.
Why is this useful?
Because NumPy stores data in continuous memory. Knowing how many bytes each element takes helps understand memory usage.
✅ 2. What is size in NumPy?
size tells you the total number of elements in the array.
✅ Why do we use them?
✅ itemsize is used to:
Check memory usage
Optimize large datasets
Know datatype storage size
✅ size is used to:
Know how many elements you have
Loop through arrays
Validate shapes
Do operations like reshape

In [30]:
arr=np.array([1,2,3,4,5],dtype="int64")
print(arr.size)#it tell the number of element in a array
print(arr.itemsize)#it tell the size of element in a array
#int32 takes 4 bytes int64 takes 8 bytes of memory

5
8


✅ Difference between size and itemsize
Function	Meaning	                          Example Output
size	    Total number of values in array	  6
itemsize	Bytes used by each value	      4

In [31]:
arr=np.array(["a","b"])

In [32]:
arr

array(['a', 'b'], dtype='<U1')

In [33]:
print(arr.size)#it tell number of elements
print(arr.itemsize) # it tell the size of element

2
4


### dtype 

In [34]:
print(a1.dtype)
print(a2.dtype)
print(a3.dtype)

int32
float64
int32


### Changing Datatype

✅ What is astype() in NumPy?
astype() is used to change the data type (dtype) of a NumPy array.
It creates a new array with the new datatype.
The original array does NOT change unless you assign it back.

✅ Why do we need astype()?
We use it when:
✅ You want to convert values to a different type
int → float
float → int
string → int
bool → int
etc.
✅ You want to reduce memory
Example: converting int64 to int8 to save memory.
✅ You want correct type for ML/DL models
Many ML models need data in float32 or int32.
✅ You want to avoid errors in operations
Example: dividing two int arrays results in float — you may force it.

✅ Important Points
astype() always returns a NEW array.
If conversion is impossible → error
Example: "abc".astype(int) ❌
Used a lot in Machine Learning, Data Cleaning, and Memory Optimization.

In [35]:
import numpy as np
arr = np.array([1, 2, 3])
new_arr = arr.astype(float)
print(new_arr)
print(new_arr.dtype) #The original array does NOT change unless you assign it back.

[1. 2. 3.]
float64


In [36]:
import  numpy as np 
arr_1=np.array([10,11,12,13,14])
arr_1=arr.astype(float)#Permanent dtype change

In [37]:
arr_1.dtype

dtype('float64')

### Array Operation

✅ 1. Arithmetic Operations (Element-wise)
These work on every element of the array.

In [38]:
a1=np.arange(12).reshape(3,4)
a2=np.arange(12,24).reshape(3,4)

In [39]:
print("a1+a2 = ",a1 + a2)
print("\n")
print("a1-a2 = ",a1 - a2)
print("\n")  
print("a1*a2 = ",a1 * a2)
print("\n")  
print("a1/a2 = ",a1 / a2)
print("\n")    
print("a1%/a2 = ",a1 % a2)
print("\n")
print(a1 ** 2) 

a1+a2 =  [[12 14 16 18]
 [20 22 24 26]
 [28 30 32 34]]


a1-a2 =  [[-12 -12 -12 -12]
 [-12 -12 -12 -12]
 [-12 -12 -12 -12]]


a1*a2 =  [[  0  13  28  45]
 [ 64  85 108 133]
 [160 189 220 253]]


a1/a2 =  [[0.         0.07692308 0.14285714 0.2       ]
 [0.25       0.29411765 0.33333333 0.36842105]
 [0.4        0.42857143 0.45454545 0.47826087]]


a1%/a2 =  [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


[[  0   1   4   9]
 [ 16  25  36  49]
 [ 64  81 100 121]]


✅ 2. Comparison Operations (Element-wise)

In [40]:
print(a1>a2,"\n")
print(a1==a2,"\n")
print(a1<=a2,"\n")
print(a1>=a2,"\n")
print(a1!=a2,"\n")

[[False False False False]
 [False False False False]
 [False False False False]] 

[[False False False False]
 [False False False False]
 [False False False False]] 

[[ True  True  True  True]
 [ True  True  True  True]
 [ True  True  True  True]] 

[[False False False False]
 [False False False False]
 [False False False False]] 

[[ True  True  True  True]
 [ True  True  True  True]
 [ True  True  True  True]] 



✅ 3. Aggregate Functions
Used to get summary values.

✅ 1. arr.sum()
Adds all elements in the array.
✅ 2. arr.min()
Returns the smallest value in the array.
✅ 3. arr.max()
Returns the largest value in the array.
✅ 4. arr.mean()
Returns the average.
✅ 5. arr.std()
Returns standard deviation, which tells how spread out the numbers are.
Low std → values are close to mean
High std → values are far from mean
✅ 6. arr.prod()
Returns the product of all values (multiplication of all elements).
✅ Use when you need multiplication of all elements.

In [41]:
arr=np.array([10,20,30,40,50,60,70])
print(arr.sum())
print(arr.max())
print(arr.min())
print(arr.mean())
print(arr.std())
print(arr.prod())

280
70
10
40.0
20.0
-1139607552


In [42]:
arr_2=np.array([[1,2,3],[4,5,6]])
print(arr_2)
print(np.max(arr_2,axis=0)) #it show greatest number from colunms

[[1 2 3]
 [4 5 6]]
[4 5 6]


In [43]:
print(np.max(arr_2,axis=1)) #greatest numbers from row 

[3 6]


In [44]:
print(np.median(arr_2,axis=0))
print(np.median(arr_2,axis=1))

[2.5 3.5 4.5]
[2. 5.]


In [45]:
print(np.var(arr_2,axis=0))
print(np.var(arr_2,axis=1))

[2.25 2.25 2.25]
[0.66666667 0.66666667]


✅ 4. Array-wise Operations in NumPy
Array-wise operations mean performing operations on entire arrays at once — element by element — without using loops.
This is one of the biggest advantages of NumPy.

✅ 1. Dot Product (Vector Dot Product)
Dot product is used between two 1D arrays (vectors).
✅ 2. Matrix Multiplication (@)
Matrix multiplication is NOT element-wise.
✅ Requirement:
Columns of A must equal rows of B.

In [46]:
a=np.array([1,2,3,4,5,6])
b=np.array([7,8,9,10,11,12])
print(np.dot(a,b))

217


#### formula
[a1, a2, a3] · [b1, b2, b3]  
= a1*b1 + a2*b2 + a3*b3
#### why
1*4 + 2*5 + 3*6  
= 4 + 10 + 18  
= 32


In [47]:
A = np.array([[1,2,3],
              [4,5,6]])
B = np.array([[7,8],
              [9,10],
              [11,12]])
print(A @ B)        # matrix multiplication
print(np.matmul(A, B))  # same result

[[ 58  64]
 [139 154]]
[[ 58  64]
 [139 154]]


✅ Why?
For position (row1, col1):
1*7 + 2*9 + 3*11
= 7 + 18 + 33
= 58

### ✅ 5. Broadcasting (very important)

👉 If two arrays don’t have the same size, NumPy automatically repeats the smaller one
so the operation can still work.
✅ No loop
✅ No manual copying
✅ NumPy does it automatically
✅ No need to write loops
✅ No need to manually resize arrays
if you have an array of this number[1, 2, 3, 4]
and you want to add 5:
[1, 2, 3, 4]  
+ 5
NumPy changes the 5 to:
[5, 5, 5, 5]
Automatically!
✅ Broadcasting – SUPER SHORT SUMMARY
✅ NumPy can stretch arrays with dimension 1
✅ It stretches them to match the bigger array
✅ Makes array operations very fast
✅ Avoids loops automatically

In [48]:
print(arr_1)
print(arr_1+5) # 5 is add to each element

[1. 2. 3.]
[6. 7. 8.]


In [49]:
A=np.array([[1,2,3],[4,5,6]])
B=np.array([10,20,30])
print(A+B)

[[11 22 33]
 [14 25 36]]


NumPy makes b like this:
[10,20,30]
[10,20,30]
Then adds them.

In [50]:
B=np.array([2,2,2])
print(A*B)#👉 NumPy stretches b to match A.

[[ 2  4  6]
 [ 8 10 12]]


### ✅ Relational Operators in Arrays

Relational operators compare values.
You already know these:
Operator	Meaning
>	        greater than
<	        less than
>=	        greater or equal
<=	        less or equal
==	        equal
!=	        not equal
✅ In NumPy, relational operators work element-wise
It compares every element one by one and returns a boolean array.

In [51]:
print(arr)
arr>20

[10 20 30 40 50 60 70]


array([False, False,  True,  True,  True,  True,  True])

In [52]:
arr==20

array([False,  True, False, False, False, False, False])

In [53]:
arr!=30

array([ True,  True, False,  True,  True,  True,  True])

### ✅ Trigonometric Functions 

✅ Trigonometric Functions in NumPy (Very Easy Explanation)
NumPy lets you do math with angles using simple functions.

➤ sin()
Returns the sine value.
➤ cos()
Returns the cosine value.
➤ tan()
Returns the tangent value.

In [54]:
print(np.sin(0))
print(np.sin(1))

0.0
0.8414709848078965


In [55]:
print(np.cos(0))
print(np.cos(1))

1.0
0.5403023058681398


In [56]:
print(np.tan(0))
print(np.tan(1))

0.0
1.5574077246549023


✅ Important Note: NumPy uses radians, not degrees!

In [57]:
np.sin(90) #This is wrong, because 90 is treated as 90 radians, not 90°.

0.8939966636005579

In [58]:
np.sin(np.deg2rad(90)) #it helps us to convert degree to radians 

1.0

✅ 2. Convert Degrees ↔ Radians
➤ deg → rad

In [59]:
np.deg2rad(90)

1.5707963267948966

➤ rad → deg

In [60]:
np.rad2deg(np.pi/2) 

90.0

✅ 3. Inverse Trigonometric Functions (Return Angle)

Inverse trigonometric functions give the angle when you already know the sin, cos, or tan value.
⚠️ Important:
These functions return the angle in radians, not degrees.

✅ Example 1: arcsin (inverse of sin)

In [61]:
np.arcsin(1)

1.5707963267948966

In [62]:
#This value = π/2 rad = 90°
np.degrees(np.arcsin(1))

90.0

✅ Example 2: arccos (inverse of cos)

In [63]:
np.arccos(0)

1.5707963267948966

Convert to degrees:

In [64]:
np.degrees(np.arccos(0))

90.0

✅ Example 3: arctan (inverse of tan)

In [65]:
np.arctan(1)

0.7853981633974483

Convert to degrees:

In [66]:
np.degrees(np.arctan(1))

45.0

✅ 1. Exponential Functions (eˣ)
These functions deal with exponentials.
✅ np.exp(x)
Gives eˣ (e = 2.718…).

In [67]:
np.exp(4)

54.598150033144236

In [68]:
np.exp(2)

7.38905609893065

✅ np.exp2(x)
Gives 2ˣ

In [69]:
np.exp2(2)

4.0

In [70]:
np.exp2(3)

8.0

✅ np.expm1(x)
Gives eˣ – 1
✔ More accurate for very small x.

In [71]:
np.expm1(0.0001)

0.00010000500016667085

In [72]:
np.expm1(0.00009)

9.000405012150273e-05

✅ 2. Logarithmic Functions
✅ np.log(x)
Natural log (base e).

In [73]:
np.log(10)

2.302585092994046

✅ np.log10(x)

Log base 10

In [74]:
np.log10(100)

2.0

✅ np.log2(x)

Log base 2

In [75]:
np.log2(8)


3.0

In [76]:
np.log2(4)

2.0

✅ np.log1p(x)

Log of (1 + x)

In [77]:
np.log1p(4)

1.6094379124341003

✅ 3. Why do we need log functions?

✅ Used in ML, Stats, Data Science:

Normalization

Probability (log probability)

Avoid overflow errors

Feature transformation

Example: Log transform makes big numbers smaller:

In [78]:
arr = np.array([1, 10, 100, 1000])
np.log10(arr)

array([0., 1., 2., 3.])

✅ np.round() — Round numbers to nearest value

In [79]:
np.round(3.567)

4.0

In [80]:
np.round(np.random.random((2,3)))#this create a array of random number between 0 and 1 (2 rows and 3 colunms)

array([[0., 0., 1.],
       [0., 1., 0.]])

In [81]:
np.round(np.random.random((2,3))*100)#this multiple 100 with all the element which are genrated randomlly

array([[ 9., 45., 54.],
       [94., 13., 54.]])

✅ Round to a specific number of decimal places
✅ Example: round to 2 decimal places

In [82]:
np.round(1.7899,2)

1.79

✅ Round to 1 decimal place

In [83]:
np.round(5.678, 1)

5.7

✅ Round an entire array

In [84]:
arr_12=np.array([2.237,8.989,1.223,5.978])
np.round(arr_12,2)

array([2.24, 8.99, 1.22, 5.98])

✅ Round to nearest 10, 100, etc.

✅ Round to nearest 10 (use negative decimals)

In [85]:
np.round(123, -1)

120

✅ Round to nearest 100

In [86]:
np.round(1234,-2)

1200

✅ 1. np.floor() → Always round DOWN
✔ Meaning:

floor() takes a number and rounds it DOWN to the nearest whole number.

In [87]:
print(np.floor(4.898))
print(np.floor(9.233))

4.0
9.0


In [88]:
print(np.floor(-2.34))

-3.0


✅ Why -3?

Because floor always goes down, and for negative numbers:

DOWN from -2.3 = -3

✅ 2. np.ceil() → Always round UP
✔ Meaning:

ceil() takes a number and rounds it UP to the nearest whole number.

In [89]:
print(np.ceil(6.7982))
print(np.ceil(2.6790))
print(np.ceil(-2.34))

7.0
3.0
-2.0


✅ Why -2?

Because ceil always goes UP, and for negative numbers:

UP from -2.3 = -2

In [90]:
arr_100=np.linspace(1,5,10)
print(arr_100)
print(np.ceil(arr_100))
print(np.floor(arr_100))
print(arr_100)

[1.         1.44444444 1.88888889 2.33333333 2.77777778 3.22222222
 3.66666667 4.11111111 4.55555556 5.        ]
[1. 2. 2. 3. 3. 4. 4. 5. 5. 5.]
[1. 1. 1. 2. 2. 3. 3. 4. 4. 5.]
[1.         1.44444444 1.88888889 2.33333333 2.77777778 3.22222222
 3.66666667 4.11111111 4.55555556 5.        ]


Indexing and Slicing

In [91]:
a1_=np.arange(1,13)
a2_=np.arange(13,25).reshape(3,4)
a3_=np.arange(25,37).reshape(3,2,2)
print(a1_)
print(a2_)
print(a3_)

[ 1  2  3  4  5  6  7  8  9 10 11 12]
[[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]
[[[25 26]
  [27 28]]

 [[29 30]
  [31 32]]

 [[33 34]
  [35 36]]]


✅ 1️⃣ Indexing & Slicing in 1D Array

In [92]:
print(a1_[0])
print(a1_[-1])#last element of array
print(a1_[6])

1
12
7


✅ Slicing (select a range)

In [93]:
print(a1_[:3])#it prints from 0 to 2 index exclude 3 
print(a1_[::2])#it will skip every 2 element 
print(a1_[::])#it will print all the element
print(a1_[1:8])#it will print all the elements from 1 to 7

[1 2 3]
[ 1  3  5  7  9 11]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[2 3 4 5 6 7 8]


✅ 2️⃣ Indexing & Slicing in 2D Array

✅ Indexing
✔ Access single element

In [94]:
print(a2_)

[[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]


In [95]:
print(a2_[2][1]) # 2 row and 1 colums

22


In [96]:
print(a2_[0,3])# 0 row and 3 colums

16


In [97]:
print(a2_[2][2]) #2 row and 2 colums
print(a2_[2,2])#2 row and 2 colums

23
23


In [98]:
print(a2_[0][0])
print(a2_[0,0]) 

13
13


✅ Slicing rows

In [99]:
print(a2_)

[[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]


In [100]:
print(a2_[0])

[13 14 15 16]


In [101]:
print(a2_[::2])

[[13 14 15 16]
 [21 22 23 24]]


In [102]:
print(a2_[1:])

[[17 18 19 20]
 [21 22 23 24]]


✅ Slicing columns

In [103]:
print(a2_[:,0])

[13 17 21]


In [104]:
print(a2_[:,2])

[15 19 23]


In [105]:
print(a2_[:,3])

[16 20 24]


In [106]:
print(a2_[:,:]) #it print all rows and colums

[[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]


✅ Slicing submatrix

In [107]:
a2_[:2,1:3]#This means: 0 to 1 rows and 1 to 2 columns

array([[14, 15],
       [18, 19]])

In [108]:
a2_

array([[13, 14, 15, 16],
       [17, 18, 19, 20],
       [21, 22, 23, 24]])

In [109]:
a2_[:2:,1::]

array([[14, 15, 16],
       [18, 19, 20]])

In [110]:
a2_[1::2,::3]

array([[17, 20]])

In [111]:
a2_[::2,::3]

array([[13, 16],
       [21, 24]])

In [112]:
a2_[1:,2:] #this means: 1 to end row and 2 to end columns

array([[19, 20],
       [23, 24]])

In [113]:
a2_[1:,1:3] # this means :1 to end rows and 1 to 2 columns

array([[18, 19],
       [22, 23]])

In [114]:
a2_[:2,:2] #this means : start to 1 rows and start to 1 columns

array([[13, 14],
       [17, 18]])

In [115]:
a2_[1:,:2] #this means : 1 to end rows and start to 1 columns

array([[17, 18],
       [21, 22]])

✅ 3️⃣ Indexing & Slicing in 3D Array

In [116]:
a3_

array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

✅ Indexing
✔ Access element:

In [117]:
a3_[0,1,1] # 0 dim 1 row and 1 columns

28

In [118]:
a3_[2,1,0] # 2 dim 1 row and 1 columns 

35

✅ Slicing one block

In [119]:
a3_[0] #it gives us 0 dim

array([[25, 26],
       [27, 28]])

In [120]:
a3_[1] #it gives us 1 dim

array([[29, 30],
       [31, 32]])

✅ Slicing rows from all blocks

In [121]:
a3_[:,1] #it has print the 1 rows of all the dim 

array([[27, 28],
       [31, 32],
       [35, 36]])

In [122]:
a3_

array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

In [123]:
a3_[0,::2,::]

array([[25, 26]])

In [124]:
a3_[2,::,1:0:-1]

array([[34],
       [36]])

In [125]:
a3_[1,::,:1:]

array([[29],
       [31]])

In [126]:
a3_[2,0:2,1:2]

array([[34],
       [36]])

✅ Slicing specific columns

In [127]:
a3_[:,:,1] #it have print 1 columns from every dim

array([[26, 28],
       [30, 32],
       [34, 36]])

✅ What is Iteration in NumPy?

Iteration means going through every element of an array one by one using loops.

1D array iteration

In [128]:
a1_

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [129]:
for i in a1_:
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12


Iteration on 2D Array

In [130]:
a2_

array([[13, 14, 15, 16],
       [17, 18, 19, 20],
       [21, 22, 23, 24]])

In [131]:
for i in a2_:
    print(i) #it prints rows

[13 14 15 16]
[17 18 19 20]
[21 22 23 24]


In [132]:
for i in a2_:
    for j in i:
        print(j) # it shows every element

13
14
15
16
17
18
19
20
21
22
23
24


3D array iteration

In [133]:
a3_

array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

In [134]:
for i in a3_:
    print(i)# it print dim

[[25 26]
 [27 28]]
[[29 30]
 [31 32]]
[[33 34]
 [35 36]]


In [135]:
for i in a3_:
    for j in i:
        print(j)# it prints all the rows

[25 26]
[27 28]
[29 30]
[31 32]
[33 34]
[35 36]


In [136]:
for i in a3_:
    for j in i:
        for k in j:
            print(k) #it prints all the elements

25
26
27
28
29
30
31
32
33
34
35
36


Using np.nditer() (IMPORTANT)

nditer() lets you loop through every element easily without nested loops.

In [137]:
for i in np.nditer(a2_):
    print(i)#2D array

13
14
15
16
17
18
19
20
21
22
23
24


In [138]:
for j in np.nditer(a3_):
    print(j)

25
26
27
28
29
30
31
32
33
34
35
36


Iterating with index — np.ndenumerate()

🔹 What is np.ndenumerate()?

It returns an iterator that yields pairs (index, value) for every element in the array.

index is a tuple giving the position (one element per dimension).

value is the element at that position.

It’s like doing for idx, val in enumerate(flattened): but ndenumerate gives you the multi-dimensional index.

In [139]:
a1_

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [140]:
for idx,val in np.ndenumerate(a1_):
    print(idx,val)

(0,) 1
(1,) 2
(2,) 3
(3,) 4
(4,) 5
(5,) 6
(6,) 7
(7,) 8
(8,) 9
(9,) 10
(10,) 11
(11,) 12


In [141]:
a2_

array([[13, 14, 15, 16],
       [17, 18, 19, 20],
       [21, 22, 23, 24]])

In [142]:
for idx,val in np.ndenumerate(a2_):
    print(idx,val)

(0, 0) 13
(0, 1) 14
(0, 2) 15
(0, 3) 16
(1, 0) 17
(1, 1) 18
(1, 2) 19
(1, 3) 20
(2, 0) 21
(2, 1) 22
(2, 2) 23
(2, 3) 24


In [143]:
a3_

array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

In [144]:
for idx,val in np.ndenumerate(a3_):
    print(idx,val)

(0, 0, 0) 25
(0, 0, 1) 26
(0, 1, 0) 27
(0, 1, 1) 28
(1, 0, 0) 29
(1, 0, 1) 30
(1, 1, 0) 31
(1, 1, 1) 32
(2, 0, 0) 33
(2, 0, 1) 34
(2, 1, 0) 35
(2, 1, 1) 36


✅  Transpose
Meaning:

Transpose means flip rows into columns and columns into rows.

In [145]:
arr_1.T

array([1., 2., 3.])

In [146]:
a2_.T

array([[13, 17, 21],
       [14, 18, 22],
       [15, 19, 23],
       [16, 20, 24]])

In [147]:
a3_.T

array([[[25, 29, 33],
        [27, 31, 35]],

       [[26, 30, 34],
        [28, 32, 36]]])

In [148]:
a3_

array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

✅ What is ravel() in NumPy?

ravel() converts any array (2D, 3D, etc.) into a 1-dimensional array.

👉 It does NOT create a copy (whenever possible).
👉 It gives a view of the original array.
👉 If you change the ravel output, the original array may also change.

✅ Why is it called ravel?

“Ravel” means to untangle — it untangles the array into a single long list.

In [149]:
a2_.ravel()

array([13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24])

In [150]:
a3_.ravel()

array([25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36])

✅ Does ravel() modify the original array?
Yes, it can.

Because it returns a view (same underlying data).

In [151]:
a = np.array([[10, 20],
              [30, 40]])

b = a.ravel()
b[0] = 999
print(a)

[[999  20]
 [ 30  40]]


✅ Changing b changed a also → because they share the same data.

(Only if NumPy cannot create a view, then it creates a copy.)

✅ When to use ravel()?

Use ravel() when:

✅ You just want a fast 1-D version of an array
✅ You’re doing operations like:

Iteration

Mathematical operations

Passing data into ML models

Flattening before matrix operations

✅ What is flatten() in NumPy?

flatten() converts any array (2D, 3D, etc.) into a 1-dimensional array, just like ravel().

✅ BUT flatten() always returns a COPY
❌ Changes to the flattened array do NOT affect the original array.

In [152]:
a = np.array([[1, 2],
              [3, 4]])

b = a.flatten()
print(b)

[1 2 3 4]


✅ Use flatten() when:

You do NOT want original array to be affected

Safety is more important than speed

You need a guaranteed copy

✅ Use ravel() when:

You want speed

Memory efficiency matters

You don’t mind that changes affect original array

✅ Summary (Easy Language)

ravel():
➡ makes array 1D
➡ fast
➡ does not copy
➡ changing the result changes original array

flatten():
➡ makes array 1D
➡ makes a new copy
➡ changing the result does NOT affect original

✅ What is Stacking?

Stacking means joining two or more arrays either:

✅ Vertically (top–bottom)
✅ Horizontally (left–right)

✅ 1. vstack() → Vertical Stacking

Think of vstack = vertical stack = top of each other
(Like stacking books on top of each other)

In [153]:
a=np.arange(1,6)
b=np.arange(6,11)
print(np.vstack((a,b)))
#✅ a becomes first row
#✅ b becomes second row

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]]


✅ 2. hstack() → Horizontal Stacking

Think of hstack = horizontal stack = side by side
(Like joining two benches left-right)

In [154]:
print(np.hstack((a,b)))#✅ a and b get joined side by side

[ 1  2  3  4  5  6  7  8  9 10]


In [155]:
print(np.hstack((a2_,a2_)))
print(np.vstack((a2_,a2_)))

[[13 14 15 16 13 14 15 16]
 [17 18 19 20 17 18 19 20]
 [21 22 23 24 21 22 23 24]]
[[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]
 [13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]


In [156]:
a3_


array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

In [157]:
print(np.vstack((a3_,a3_)))


[[[25 26]
  [27 28]]

 [[29 30]
  [31 32]]

 [[33 34]
  [35 36]]

 [[25 26]
  [27 28]]

 [[29 30]
  [31 32]]

 [[33 34]
  [35 36]]]


In [158]:
print(np.hstack((a3_,a3_)))

[[[25 26]
  [27 28]
  [25 26]
  [27 28]]

 [[29 30]
  [31 32]
  [29 30]
  [31 32]]

 [[33 34]
  [35 36]
  [33 34]
  [35 36]]]


Splitting arrays — hsplit, vsplit (detailed, simple)

Splitting means breaking one big array into smaller arrays.
NumPy gives special functions for common splits: hsplit (split columns) and vsplit (split rows). I’ll explain how they work, show examples, common errors, and alternatives.

1) Quick rules (easy)

vsplit → split vertically = cut top → bottom (breaks along rows)
Equivalent to np.split(..., axis=0).

hsplit → split horizontally = cut left → right (breaks along columns)
Equivalent to np.split(..., axis=1) for 2-D arrays.

Both expect the array to have enough dimensions (e.g., vsplit/hsplit are meaningful for 2D+ arrays).

By default they require equal parts. If equal division is not possible you get ValueError.

Use np.array_split to allow unequal parts.

In [159]:
a2_
print(np.hsplit(a2_,2))

[array([[13, 14],
       [17, 18],
       [21, 22]]), array([[15, 16],
       [19, 20],
       [23, 24]])]


In [160]:
a3_
print(np.hsplit(a3_,2))

[array([[[25, 26]],

       [[29, 30]],

       [[33, 34]]]), array([[[27, 28]],

       [[31, 32]],

       [[35, 36]]])]


In [161]:
import numpy as np
A = np.arange(1, 13).reshape(6, 2)  
 # shape (6,2)
A

array([[ 1,  2],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [ 9, 10],
       [11, 12]])

In [162]:
parts = np.vsplit(A, 3)
parts

[array([[1, 2],
        [3, 4]]),
 array([[5, 6],
        [7, 8]]),
 array([[ 9, 10],
        [11, 12]])]

In [163]:
B = np.arange(1, 13).reshape(3, 4) 
B

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [164]:
left, right = np.hsplit(B, 2) 
print(left)
print(right)

[[ 1  2]
 [ 5  6]
 [ 9 10]]
[[ 3  4]
 [ 7  8]
 [11 12]]


In [165]:
a3_

array([[[25, 26],
        [27, 28]],

       [[29, 30],
        [31, 32]],

       [[33, 34],
        [35, 36]]])

In [166]:
left,right=np.hsplit(a3_,2)
print(left)
print(right)

[[[25 26]]

 [[29 30]]

 [[33 34]]]
[[[27 28]]

 [[31 32]]

 [[35 36]]]


In [167]:
arr = np.arange(1,13).reshape(3,4)  # shape (3,4)
parts = np.hsplit(arr, [1, 3])
print(parts)

[array([[1],
       [5],
       [9]]), array([[ 2,  3],
       [ 6,  7],
       [10, 11]]), array([[ 4],
       [ 8],
       [12]])]


In [168]:
arr

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

# Numpy Tricks

### np.sort

✅ NumPy sort() — Simple Explanation

np.sort() is used to sort the elements of a NumPy array.

In [169]:
a=np.random.randint(1,100,15)
print(a)
np.sort(a)

[57 52 77 20  6 64 74 38 34 58 12 19 84  5 73]


array([ 5,  6, 12, 19, 20, 34, 38, 52, 57, 58, 64, 73, 74, 77, 84])

In [170]:
np.sort(a)[::-1]#it sort everything in decending order 

array([84, 77, 74, 73, 64, 58, 57, 52, 38, 34, 20, 19, 12,  6,  5])

In [171]:
b=np.random.randint(1,100,24).reshape(6,4)
print(b)
np.sort(b) #it sort each row in am array 

[[36 54 84 42]
 [70 14 43  2]
 [25 40 68 28]
 [10  4 90 32]
 [33 73 94 26]
 [17 72 57 79]]


array([[36, 42, 54, 84],
       [ 2, 14, 43, 70],
       [25, 28, 40, 68],
       [ 4, 10, 32, 90],
       [26, 33, 73, 94],
       [17, 57, 72, 79]])

In [172]:
np.sort(b,axis=0)#it sort columns wise 

array([[10,  4, 43,  2],
       [17, 14, 57, 26],
       [25, 40, 68, 28],
       [33, 54, 84, 32],
       [36, 72, 90, 42],
       [70, 73, 94, 79]])

In [173]:
np.sort(b,axis=1)#it sort row wise 

array([[36, 42, 54, 84],
       [ 2, 14, 43, 70],
       [25, 28, 40, 68],
       [ 4, 10, 32, 90],
       [26, 33, 73, 94],
       [17, 57, 72, 79]])

✅ Sort and get the indices (argsort)

If you want positions instead of values:

In [174]:
np.argsort(a) #array([[36 33 30 51 52 59 92 18 74 90  5 97 38 55 35]]) actual array
#array([ 5, 18, 30, 33, 35, 36, 38, 51, 52, 55, 59, 74, 90, 92, 97]) acending order array
#output show the index of acending order array 

array([13,  4, 10, 11,  3,  8,  7,  1,  0,  9,  5, 14,  6,  2, 12],
      dtype=int64)

✅ In-place sorting (ndarray.sort())

This sorts the array without making a new copy:

In [175]:
a.sort()
print(a)

[ 5  6 12 19 20 34 38 52 57 58 64 73 74 77 84]


In [176]:
a.sort()
a=a[::-1]
print(a)

[84 77 74 73 64 58 57 52 38 34 20 19 12  6  5]


✅ What np.append() Does (in simple words)

👉 np.append() adds extra values to an array and returns a NEW array.
It does not change the original array.

✅ Basic Syntax
np.append(array, values, axis=None)

✅ If axis=None (default)

The array is flattened into 1D

Values are appended at the end

In [177]:
np.append(a,200)

array([ 84,  77,  74,  73,  64,  58,  57,  52,  38,  34,  20,  19,  12,
         6,   5, 200])

In [178]:
a

array([84, 77, 74, 73, 64, 58, 57, 52, 38, 34, 20, 19, 12,  6,  5])

In [179]:
np.append(b,np.ones((b.shape[0],1)))

array([36., 54., 84., 42., 70., 14., 43.,  2., 25., 40., 68., 28., 10.,
        4., 90., 32., 33., 73., 94., 26., 17., 72., 57., 79.,  1.,  1.,
        1.,  1.,  1.,  1.])

In [180]:
print(b)

[[36 54 84 42]
 [70 14 43  2]
 [25 40 68 28]
 [10  4 90 32]
 [33 73 94 26]
 [17 72 57 79]]


In [181]:
new_row=np.array([[1,2,3,4]])
np.append(b,new_row,axis=0)

array([[36, 54, 84, 42],
       [70, 14, 43,  2],
       [25, 40, 68, 28],
       [10,  4, 90, 32],
       [33, 73, 94, 26],
       [17, 72, 57, 79],
       [ 1,  2,  3,  4]])

In [182]:
new_Columns=np.array([[11]
                      ,[12]
                      ,[13]
                      ,[14]
                      ,[15]
                      ,[16]])
np.append(b,new_Columns,axis=1)

array([[36, 54, 84, 42, 11],
       [70, 14, 43,  2, 12],
       [25, 40, 68, 28, 13],
       [10,  4, 90, 32, 14],
       [33, 73, 94, 26, 15],
       [17, 72, 57, 79, 16]])

In [183]:
np.append([[1,2],[3,4]], [5,6])

array([1, 2, 3, 4, 5, 6])

✅ 2. np.ones((b.shape[0], 1))

This creates a column of 1’s with same number of rows as b.

In [184]:
np.append(b,np.ones((b.shape[0],1)),axis=1)

array([[36., 54., 84., 42.,  1.],
       [70., 14., 43.,  2.,  1.],
       [25., 40., 68., 28.,  1.],
       [10.,  4., 90., 32.,  1.],
       [33., 73., 94., 26.,  1.],
       [17., 72., 57., 79.,  1.]])

In [185]:
np.append(b,np.random.random((b.shape[0],1)),axis=1) # it add a columns which have random values 

array([[36.        , 54.        , 84.        , 42.        ,  0.97998511],
       [70.        , 14.        , 43.        ,  2.        ,  0.58285343],
       [25.        , 40.        , 68.        , 28.        ,  0.93312228],
       [10.        ,  4.        , 90.        , 32.        ,  0.61794955],
       [33.        , 73.        , 94.        , 26.        ,  0.86246137],
       [17.        , 72.        , 57.        , 79.        ,  0.79847547]])

✅ What is np.concatenate()?

👉 np.concatenate() joins (merges) two or more NumPy arrays

👉 It does NOT flatten the arrays (unlike np.append())

Think of it like stitching arrays side-by-side or top-to-bottom.

✅ Basic Syntax
np.concatenate((array1, array2), axis=0)

You must pass the arrays inside a tuple ( )

You must specify axis for multi-dimensional arrays

In [186]:
_1D_array=np.arange(6)
_1D_array2=np.arange(6,12)
print(_1D_array)
print(_1D_array2)
np.concatenate((_1D_array,_1D_array2))

[0 1 2 3 4 5]
[ 6  7  8  9 10 11]


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [187]:
c=np.arange(6).reshape(2,3)
d=np.arange(6,12).reshape(2,3)

print(c)
print(d)

[[0 1 2]
 [3 4 5]]
[[ 6  7  8]
 [ 9 10 11]]


In [188]:
np.concatenate((c,d),axis=0) #it adds row 

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

In [189]:
np.concatenate((c,d),axis=1) #it add columns 

array([[ 0,  1,  2,  6,  7,  8],
       [ 3,  4,  5,  9, 10, 11]])

✅ What np.unique() Does

np.unique() finds all the unique (non-repeated) values in a NumPy array.

It also sorts the result automatically.

👉 All duplicates are removed
👉 Sorted automatically

In [190]:
import numpy as np

a = np.array([1, 3, 3, 2, 1, 5])

print(np.unique(a))

[1 2 3 5]


✅ Useful Options

1️⃣ Return counts of each unique value

In [191]:
values, counts = np.unique(a, return_counts=True)
#values = unique elements
#counts = how many times each appears
print(values,counts)

[1 2 3 5] [2 1 2 1]


2️⃣ Return index of first occurrence

In [192]:
values, index = np.unique(a, return_index=True)
print(values,index) # index tell the index of first occuring 

[1 2 3 5] [0 3 1 5]


3️⃣ Return inverse mapping

In [193]:
values, inverse = np.unique(a, return_inverse=True)#[1, 3, 3, 2, 1, 5]
print(values)
print(inverse) #Inverse tells how to reconstruct the original array.

[1 2 3 5]
[0 2 2 1 0 3]


✅ What np.expand_dims() does (in one line)

It adds a new axis (dimension) to your array.

✅ Why do we need it?

Sometimes your array shape is not compatible for:

broadcasting

stacking

machine learning models

reshaping

So we add a dimension of size 1.

✅ Syntax
np.expand_dims(arr, axis)


arr → your array

axis → position where you want to insert the new dimension

In [194]:
a=np.arange(6)
print(a)
print(a.shape)
print(np.expand_dims(a,axis=0))
print(a) #it does not change the original one it return a expended copy 
print(np.expand_dims(a,axis=1))

[0 1 2 3 4 5]
(6,)
[[0 1 2 3 4 5]]
[0 1 2 3 4 5]
[[0]
 [1]
 [2]
 [3]
 [4]
 [5]]


In [195]:
b=np.array([[1,2],[3,4],[5,6],[7,8]])
print(b.shape)
c=np.expand_dims(b,axis=2)
print(c)
print(c.shape) # 4 dim 2 rows 1 columns

(4, 2)
[[[1]
  [2]]

 [[3]
  [4]]

 [[5]
  [6]]

 [[7]
  [8]]]
(4, 2, 1)


In [196]:
a = np.array([[1,2,3],[4,5,6]])
print(a.shape)   # (2,3)

b = np.expand_dims(a, axis=2)
print(b.shape) # 2 dim 3 rows 1 columns
b


(2, 3)
(2, 3, 1)


array([[[1],
        [2],
        [3]],

       [[4],
        [5],
        [6]]])

✅ What is np.where() ?

np.where() is used to find positions where a condition is TRUE,
or to choose values based on a condition.

In [197]:
a=np.random.randint(1,100,10)
print(a)

[22  4 56 82 32 68 25 68 95  6]


In [198]:
idx=np.where(a>50)
print(idx)

(array([2, 3, 5, 7, 8], dtype=int64),)


In [199]:
np.where(a>50,100,a)

array([ 22,   4, 100, 100,  32, 100,  25, 100, 100,   6])

In [200]:
a

array([22,  4, 56, 82, 32, 68, 25, 68, 95,  6])

In [201]:
np.where(a>50,"high","low")

array(['low', 'low', 'high', 'high', 'low', 'high', 'low', 'high', 'high',
       'low'], dtype='<U4')

In [202]:
a = np.array([1, 0, 3, 0, 5])

result = np.where(a == 0, -1, a)
print(result)

[ 1 -1  3 -1  5]


In [203]:
b = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])
loc = np.where(b > 5)
print(loc)

# Numbers greater than 5 are:
# 6 → at (row 1, col 2)
# 7 → at (row 2, col 0)
# 8 → at (row 2, col 1)
# 9 → at (row 2, col 2)
# But your matrix only checks >5, so:
# ✅ 6 → row 1, col 2
# ✅ 7 → row 2, col 0
# ✅ 8 → row 2, col 1
# ✅ 9 → row 2, col 2

(array([1, 2, 2, 2], dtype=int64), array([2, 0, 1, 2], dtype=int64))


✅ What is np.argmax?

np.argmax() returns the index (position) of the maximum value in a NumPy array.

In [204]:
a=np.array([10,50,40,89,2])
print(np.argmax(a))

3


✅ Why?

The maximum value = 80
Its index = 3

In [205]:
b=np.array([[1,2,3],[4,5,6],[7,8,9]])
print(np.argmax(b))

8


In [206]:
print(np.argmax(b,axis=1))

[2 2 2]


In [207]:
b=np.random.randint(1,100,12).reshape(4,3)
print(b)
print(np.argmax(b))
print(np.argmax(b,axis=0)) #columns
print(np.argmax(b,axis=1)) # rows 

[[45 83 99]
 [51 23 56]
 [67 52 14]
 [82 62 88]]
2
[3 0 0]
[2 2 0 2]


✅ What is np.argmin?

np.argmin() returns the index (position) of the minimum value in a NumPy array.

It works exactly like np.argmax, but finds the smallest number.

In [208]:
print(np.argmin(b))
print(np.argmin(b,axis=0))
print(np.argmin(b,axis=1))

8
[0 1 2]
[0 1 2 1]


In [209]:
print(np.argmin(a))

4


✅ What is np.cumsum?

np.cumsum() gives the cumulative sum of elements in an array.

It means:
👉 Keep adding each element to the sum of all previous elements.

In [210]:
a

array([10, 50, 40, 89,  2])

In [211]:
np.cumsum(a)

array([ 10,  60, 100, 189, 191])

In [212]:
b

array([[45, 83, 99],
       [51, 23, 56],
       [67, 52, 14],
       [82, 62, 88]])

In [213]:
np.cumsum(b)

array([ 45, 128, 227, 278, 301, 357, 424, 476, 490, 572, 634, 722])

In [214]:
np.cumsum(b,axis=1)

array([[ 45, 128, 227],
       [ 51,  74, 130],
       [ 67, 119, 133],
       [ 82, 144, 232]])

In [215]:
np.cumsum(b,axis=0)

array([[ 45,  83,  99],
       [ 96, 106, 155],
       [163, 158, 169],
       [245, 220, 257]])

np.cumprod() — Cumulative Product
👉 Meaning

It multiplies values one-by-one and keeps storing the running product.

In [216]:
a

array([10, 50, 40, 89,  2])

In [217]:
b

array([[45, 83, 99],
       [51, 23, 56],
       [67, 52, 14],
       [82, 62, 88]])

In [218]:
np.cumprod(a)

array([     10,     500,   20000, 1780000, 3560000])

In [219]:
np.cumprod(b)

array([         45,        3735,      369765,    18858015,   433734345,
       -1480680456,  -421342744,  -434986208, -1794839616, -1147960448,
        1840896256, -1209886720])

In [220]:
np.cumprod(b,axis=1)

array([[    45,   3735, 369765],
       [    51,   1173,  65688],
       [    67,   3484,  48776],
       [    82,   5084, 447392]])

In [221]:
np.cumprod(b,axis=0)

array([[      45,       83,       99],
       [    2295,     1909,     5544],
       [  153765,    99268,    77616],
       [12608730,  6154616,  6830208]])

np.random.uniform() generates random floating-point numbers that are uniformly distributed between a given range.That means every number in the range has an equal chance of being selected.

np.random.uniform(low, high, size)

Parameter	Meaning

low	    The lower boundary of the range (inclusive).

high	The upper boundary of the range (exclusive).

size	Number of random values to generate (can be int or tuple).

In [222]:
x=np.random.uniform(0,1)
print(x)

0.3600389449921294


In [223]:
x=np.random.uniform(0,1,5)
x

array([0.30879455, 0.49317564, 0.21444228, 0.94313355, 0.58200023])

In [224]:
matrix = np.random.uniform(1, 100, (3, 4))
print(matrix)

[[80.02640564 55.12365712 12.25931669  2.81042146]
 [65.18628816 81.21029266 81.24453921  4.45313429]
 [24.52371459 69.44852359 66.91764015 26.72106677]]


🎯 What is Percentile?

A percentile tells you the value below which a given percentage of data falls.

Example

50th percentile → value below which 50% of data lies (same as median).

90th percentile → value below which 90% of data lies.

25th percentile → first quartile (Q1).

75th percentile → third quartile (Q3).

np.percentile(data, p)

data → array
p → percentile number (0–100)

In [225]:
a = np.array([10, 20, 30, 40, 50,60])
np.percentile(a, 50)

35.0

In [226]:
np.percentile(a, [25, 50, 75])

array([22.5, 35. , 47.5])

In [227]:
np.percentile(a,50)#it's median of array

35.0

In [228]:
np.percentile(a,40)

30.0

In [229]:
np.percentile(b,20)

46.2

In [230]:
np.percentile(b,90)

87.5

In [231]:
b

array([[45, 83, 99],
       [51, 23, 56],
       [67, 52, 14],
       [82, 62, 88]])

📊 What is np.histogram()?

np.histogram() is used to calculate the data for a histogram:

It splits your data into bins (ranges)

It counts how many values fall into each bin

It does NOT draw the graph — it only gives the numbers needed to draw one.

In [232]:
import matplotlib.pyplot as plt

In [233]:
a

array([10, 20, 30, 40, 50, 60])

✔ Output of np.histogram()

It returns two arrays:

hist → number of values in each bin

bin_edges → the boundary (start & end) of each bin

In [234]:
hist,bins=np.histogram(a,bins=[0,10,20,30,40,50,60,70,80,90,100])
print(hist)
print(bins)

[0 1 1 1 1 1 1 0 0 0]
[  0  10  20  30  40  50  60  70  80  90 100]


In [235]:
hist,bins=np.histogram(a,bins=[0,50,100])
print(hist)
print(bins)

[4 2]
[  0  50 100]


📌 When to Use np.histogram()?

Use it when:

✔ You want the exact counts of values per bin

✔ You want numerical histogram data

✔ You are analyzing distribution

✔ You want to manually process bins before plotting

📌 What is np.corrcoef()?

np.corrcoef() calculates the correlation coefficient between variables.

👉 It tells how strongly two arrays are related.

Correlation value (r):

Value	Meaning
+1	Perfect positive relation (both increase together)

0	No relation

–1	Perfect negative relation (one increases, other decreases)

🧮 Syntax
np.corrcoef(x, y)

In [236]:
salary=np.array([20000,40000,25000,35000,60000])
experience=np.array([1,3,2,4,2])

np.corrcoef(salary,experience)

array([[1.        , 0.25344572],
       [0.25344572, 1.        ]])

In [237]:
x=np.array([10,20,30,40])
y=np.array([15,25,35,45])

print(np.corrcoef(x,y))
#✔ When one increases, the other also increases → Perfect positive (1.0)

[[1. 1.]
 [1. 1.]]


In [238]:
y=np.array([40,30,20,10])
print(np.corrcoef(x,y))
#✔ When x increases, y decreases → Perfect negative correlation (-1.0)

[[ 1. -1.]
 [-1.  1.]]


In [239]:
print(np.corrcoef(y,x))

[[ 1. -1.]
 [-1.  1.]]


In [240]:
x=np.random.randint(1,50,10)
y=np.random.randint(1,50,10)
print(np.corrcoef(x,y))

[[ 1.        -0.4755161]
 [-0.4755161  1.       ]]


✅ 1. Why is correlation almost 0 (0.00007)?

Because:

x is random numbers

y is random numbers

Random arrays have no relationship

So correlation will always be very close to 0 

✅ 2. Why the matrix shape is 2×2?

Because np.corrcoef(x, y) returns:

[[ corr(x,x), corr(x,y) ],
 [ corr(y,x), corr(y,y) ]]

 Explanation:

1 → correlation of x with itself (always 1)

1 → correlation of y with itself (always 1)

0.00007151 → correlation of x with y

0.00007151 → correlation of y with x

In [241]:
x

array([ 3, 35, 26, 15, 38, 11, 20, 34, 35, 27])

In [242]:
y

array([39,  5, 39, 11, 27, 23, 33, 12,  4, 29])

In [243]:
#🧠 Example 3: No correlation
x = np.array([1,2,3,4,5])
y = np.array([7,9,3,8,2])

print(np.corrcoef(x, y))

[[ 1.         -0.55844039]
 [-0.55844039  1.        ]]


✅ Step 1 — Understand the matrix
[[ corr(x,x)     corr(x,y) ],
 [ corr(y,x)     corr(y,y) ]]

So:

1 → correlation of x with x (always 1)

1 → correlation of y with y (always 1)

-0.55844039 → correlation of x with y

🚨 Main thing: Why is correlation negative (–0.55)?

Correlation checks:

When x increases, does y increase? → positive correlation

When x increases, does y decrease? → negative correlation

If random → near 0

Your values:

x:
1, 2, 3, 4, 5  (increasing)

y:
7, 9, 3, 8, 2  (up, up, down, up, down)


Y is going:

↑ from 7 → 9

↓ from 9 → 3

↑ from 3 → 8

↓ from 8 → 2

This means:

When x increases
→ y sometimes goes up, sometimes down
→ but more often in the opposite direction

So overall relationship = negative trend

Meaning of –0.558 correlation

Negative = as x increases, y tends to decrease

Medium strength (since -0.55 ≈ moderate negative correlation)

Not strong, not weak — in middle

In [244]:
data = np.array([[1, 2, 3],
                 [4, 5, 6]])

np.corrcoef(data)


array([[1., 1.],
       [1., 1.]])

✅ np.isin() – What it does

np.isin() checks if each element of an array is present in another array.

It returns a Boolean array (True/False).

📌 Syntax
np.isin(array, values)

array → the array you want to check

values → the elements you want to search for

In [245]:
a=np.array([10,20,30,40,50])
check=np.array([20,40])
print(np.isin(a,check))

[False  True False  True False]


Explanation

20 and 40 are present in a

So their positions become True

In [246]:
a[np.isin(a,[20,50,30])] #it show all the values 

array([20, 30, 50])

In [247]:
a[np.isin(a,[90,50])] #This extracts only values that are present 

array([50])

✅ What is np.flip?

np.flip() reverses the elements of an array along any axis.

Think of it like turning the array upside down or backwards.

✅ Syntax
np.flip(array, axis)

axis = 0 → flip rows (top ↔ bottom)

axis = 1 → flip columns (left ↔ right)

axis not given → flip entire array

In [248]:
np.flip(a)

array([50, 40, 30, 20, 10])

In [249]:
a2_

array([[13, 14, 15, 16],
       [17, 18, 19, 20],
       [21, 22, 23, 24]])

In [250]:
np.flip(a2_,axis=0) #Rows swapped (top ↔ bottom).

array([[21, 22, 23, 24],
       [17, 18, 19, 20],
       [13, 14, 15, 16]])

In [251]:
np.flip(a2_,axis=1) #Columns reversed (left ↔ right).

array([[16, 15, 14, 13],
       [20, 19, 18, 17],
       [24, 23, 22, 21]])

In [252]:
a2_ 

array([[13, 14, 15, 16],
       [17, 18, 19, 20],
       [21, 22, 23, 24]])

In [253]:
np.flip(a2_) #Entire array flipped.

array([[24, 23, 22, 21],
       [20, 19, 18, 17],
       [16, 15, 14, 13]])

✅ What is np.put?

np.put() is used to change specific elements in a NumPy array by giving:

the index positions you want to replace

the new values you want to put there

It modifies the original array.

✅ Syntax
np.put(array, indices, values)

Parameters

array → the array you want to change

indices → positions where new values should go

values → the new values to place at those positions

In [254]:
a

array([10, 20, 30, 40, 50])

In [255]:
np.put(a,[1,4],[11,400]) #it make parmenent changes 

In [256]:
a #✔ index 1 becomes 100
#✔ index 4 becomes 400

array([ 10,  11,  30,  40, 400])

In [257]:
a = np.array([1, 2, 3, 4])
np.put(a, [0, 2], [9, 8, 7]) #Explanation:
#Only first two values are used because indices are just 2. 
print(a)

[9 2 8 4]


In [258]:
a2_

array([[13, 14, 15, 16],
       [17, 18, 19, 20],
       [21, 22, 23, 24]])

In [259]:
np.put(a2_,[0,5],[55,200])
a2_ #📌 Important: np.put treats the 2D array as flattened:

array([[ 55,  14,  15,  16],
       [ 17, 200,  19,  20],
       [ 21,  22,  23,  24]])

🎯 Why use np.put()?

It is useful when:

You want to set values by flat index, not row/column.

You want to modify multiple positions at once.

You don't want to reshape array manually.

🎯 Summary (1 line)

np.put() inserts new values at specified flat indices and updates the original array.

✅ np.delete() — Full Explanation (Very Easy)

np.delete() is used to remove elements from a NumPy array.

Syntax
np.delete(arr, obj, axis=None)

Parameters

arr → your original array

obj → index (or list of indices) you want to delete

axis

None → delete from flattened array

0 → delete rows

1 → delete columns

In [260]:
a

array([9, 2, 8, 4])

In [261]:
np.delete(a,3)

array([9, 2, 8])

In [262]:
np.delete(a,[0,3])

array([2, 8])

In [263]:
a=np.random.randint(1,100,10)
print(a)

[35 43 34 45 13 73 79 86 77 70]


In [264]:
a=np.delete(a,[0,9,4]) # if return the array in veriable then only it will change permenently
#otherwise it will just return new array
print(a)

[43 34 45 73 79 86 77]


In [265]:
#a2=a.reshape(5,2)

In [266]:
a2

array([[12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [267]:
np.delete(a2,[1,5,8])

array([12, 14, 15, 16, 18, 19, 21, 22, 23])

In [268]:
np.delete(a2,1,axis=1) #it delete one columns

array([[12, 14, 15],
       [16, 18, 19],
       [20, 22, 23]])

In [269]:
#np.delete(a2,4,axis=0) #it remove one row 

In [270]:
#np.delete(a2,[0,2,3],axis=0) #it delete 0 , 2 and 3 rows for the array and return us do not make permenent changes 

### set function

✅ 1. np.unique() — Find Unique Elements (Remove duplicates)
📌 What it does:

Removes duplicates

Sorts the values

Returns only unique values

In [271]:
a = np.array([1, 2, 2, 3, 4])
b = np.array([3, 4, 4, 5, 6])

In [272]:
np.unique(a)

array([1, 2, 3, 4])

In [273]:
np.unique(a,return_counts=True) #it return the number of how many time a number is repeated

(array([1, 2, 3, 4]), array([1, 2, 1, 1], dtype=int64))

Meaning:

1 appears → 1 time

2 appears → 2 times

3 → 1 time

4 → 1 time

In [274]:
# ➤ return_index=True
# Shows where the first appearance happened.
np.unique(a, return_index=True)

(array([1, 2, 3, 4]), array([0, 1, 3, 4], dtype=int64))

In [275]:
# ➤ return_inverse=True
# Useful to rebuild original array from unique values.
vals, inv = np.unique(a, return_inverse=True)
print(vals)
print(inv)

[1 2 3 4]
[0 1 1 2 3]


Step 1 — vals (unique values)
vals = [1, 2, 3, 4]
index   0  1  2  3

These are the unique sorted values.

Indexes:

1 → index 0

2 → index 1

3 → index 2

4 → index 3

Step 2 — inv (inverse mapping)
inv = [0, 1, 1, 2, 3]

Each number in inv tells:
👉 From vals, which index to pick
to reconstruct the original array.

Step 3 — Let’s REBUILD the original array

Use:

vals[inv]

Let's decode it:

inv[0] = 0 → vals[0] = 1

inv[1] = 1 → vals[1] = 2

inv[2] = 1 → vals[1] = 2

inv[3] = 2 → vals[2] = 3

inv[4] = 3 → vals[3] = 4

So you get:

[1, 2, 2, 3, 4]

Which is your original array (a).

In [276]:
a=np.array([1,2,3,4,5,6])
b=np.array([1,4,5,7,8,9])
np.union1d(a,b)

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [277]:
np.intersect1d(a,b)

array([1, 4, 5])

In [278]:
np.setdiff1d(a,b)

array([2, 3, 6])

In [279]:
np.setxor1d(a,b)

array([2, 3, 6, 7, 8, 9])

🔷 1. np.in1d() – Check Membership (True/False)

np.in1d() tells you which elements of one array exist in another array.

It works similar to np.isin()
(but returns a 1-D Boolean array always).

✅ Syntax
np.in1d(array, test_elements)

array → the array you want to check

test_elements → values to search for

Returns → Boolean array (True/False)

In [280]:
np.in1d(a,2)

array([False,  True, False, False, False, False])

In [281]:
a[np.in1d(a,2)]

array([2])

🔷 2. np.clip() – Limit values within a range

np.clip() forces all array values to stay between a minimum and maximum.

Think of it as cutting off extreme values.

✅ Syntax
np.clip(array, min_value, max_value)


Any value less than min_value → replaced by min_value

Any value greater than max_value → replaced by max_value

In [282]:
a = np.array([5, 10, 15, 20, 25])
print(np.clip(a, 10, 20))
# ✔ Anything < 10 → becomes 10
# ✔ Anything > 20 → becomes 20

[10 10 15 20 20]


In [283]:
data = np.array([-5, -2, 0, 3, 8])
result = np.clip(data, 0, 5)
print(result)
# Negative → clipped to 0
# Value above 5 → clipped to 5

[0 0 0 3 5]


In [284]:
marks = np.array([120, 85, -10, 95])
np.clip(marks, 0, 100)

array([100,  85,   0,  95])